lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp

```diff
@@ -0,0 +1,409 @@
+
+    // Process the annotation.
+    collectAnno(inst, instAnnos[0]);
+  });
+
+  // Propagate the DUT marking to all arbitrarily nested submodules of the DUT.
+  LLVM_DEBUG(llvm::dbgs() << "Marking DUT hierarchy\n");
+  SmallVector<InstanceGraphNode *> worklist;
+  for (Operation *op : dutModules)
+    if (auto module = dyn_cast<hw::HWModuleLike>(op))
+      worklist.push_back(instanceGraph->lookup(module));
+  while (!worklist.empty()) {
+    auto module = worklist.pop_back_val();
+    LLVM_DEBUG(llvm::dbgs()
+               << "- " << module->getModule().moduleName() << "\n");
+    for (auto *instRecord : *module) {
+      auto target = instRecord->getTarget();
+      if (dutModules.insert(target->getModule()).second)
+        worklist.push_back(target);
+    }
+  }
+}
+
+/// Process an extraction annotation on an instance into a corresponding
+/// `ExtractionInfo` and a spot on the worklist for later moving things around.
+void ExtractInstancesPass::collectAnno(InstanceOp inst, Annotation anno) {
+  LLVM_DEBUG(llvm::dbgs() << "Processing instance `" << inst.name() << "` "
+                          << anno.getDict() << "\n");
+
+  auto getStringOrError = [&](StringRef member) {
+    auto attr = anno.getMember<StringAttr>(member);
+    if (!attr) {
+      inst.emitError("missing `")
+          << member << "` attribute in `" << anno.getClass() << "` annotation";
+      anyFailures = true;
+    }
+    return attr;
+  };
+
+  if (anno.isClass(extractBlackBoxAnnoClass)) {
+    auto filename = getStringOrError("filename");
+    auto prefix = getStringOrError("prefix");
+    auto dest = anno.getMember<StringAttr>("dest"); // optional
+    if (!filename || !prefix)
+      return;
+
+    // Find the parent module into which this instance should be propagate
+    // upwards. This does not include the optional wrapper module which gets
+    // created as a separate step.
+    // TODO: Is this even needed?
+
+    ExtractionInfo info;
+    info.traceFilename = filename;
+    info.prefix = prefix;
+    info.wrapperModule = (dest ? dest.getValue() : "");
+    // info.upwardsTarget = ;
+    extractionWorklist.push_back({inst, info});
+    return;
+  }
+}
+
+/// Move instances in the extraction worklist upwards in the hierarchy. This
+/// iteratively pushes instances up one level of hierarchy until they have
+/// arrived in the desired container module.
+void ExtractInstancesPass::extractInstances() {
+  // The list of ports to be added to an instance's parent module. Cleared and
+  // reused across instances.
+  SmallVector<std::pair<unsigned, PortInfo>> newPorts;
+  // The uniquified wiring prefix for each instance.
+  DenseMap<Operation *, SmallString<16>> instPrefices;
+  // The number of instances with the same prefix. Used to uniquify prefices.
+  DenseMap<StringRef, unsigned> prefixUniqueIDs;
+
+  while (!extractionWorklist.empty()) {
+    InstanceOp inst;
+    ExtractionInfo info;
+    std::tie(inst, info) = extractionWorklist.pop_back_val();
```

Comment:
The following may work:
```suggestion
    auto [inst, info] = extractionWorklist.pop_back_val();
```

Reply:
:partying_face: TIL: Fancy new syntax! Love it!